amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -299,149 +257,44 @@ private synchronized void assertHashRangeOfClosedShardIsCovered(final Shard clos
-     *         we begin processing data from any of its descendants.
-     *   * A shard does not start processing data until data from all its parents has been processed.
-     * Note, if the initial position is LATEST and a shard has two parents and only one is a descendant - we'll create
-     * leases corresponding to both the parents - the parent shard which is not a descendant will have  
-     * its checkpoint set to Latest.
-     * 
-     * We assume that if there is an existing lease for a shard, then either:
-     *   * we have previously created a lease for its parent (if it was needed), or
-     *   * the parent shard has expired.
-     * 
-     * For example:
-     * Shard structure (each level depicts a stream segment):
-     * 0 1 2 3 4   5   - shards till epoch 102
-     * \ / \ / |   |
-     *  6   7  4   5   - shards from epoch 103 - 205
-     *   \ /   |  / \
-     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
-     * Current leases: (3, 4, 5)
-     * New leases to create: (2, 6, 7, 8, 9, 10)
-     * 
-     * The leases returned are sorted by the starting sequence number - following the same order
-     * when persisting the leases in DynamoDB will ensure that we recover gracefully if we fail
-     * before creating all the leases.
-     *
-     * If a shard has no existing lease, is open, and is a descendant of a parent which is still open, we ignore it
-     * here; this happens when the list of shards is inconsistent, which could be due to pagination delay for very
-     * high shard count streams (i.e., dynamodb streams for tables with thousands of partitions).  This can only
-     * currently happen here if ignoreUnexpectedChildShards was true in syncShardleases.
-     *
-     * 
-     * @param shards List of all shards in Kinesis (we'll create new leases based on this set)
-     * @param currentLeases List of current leases
-     * @param initialPosition One of LATEST, TRIM_HORIZON, or AT_TIMESTAMP. We'll start fetching records from that
-     *        location in the shard (when an application starts up for the first time - and there are no checkpoints).
-     * @param inconsistentShardIds Set of child shard ids having open parents.
-     * @return List of new leases to create sorted by starting sequenceNumber of the corresponding shard
-     */
-    static List<Lease> determineNewLeasesToCreate(final List<Shard> shards, final List<Lease> currentLeases,
-            final InitialPositionInStreamExtended initialPosition, final Set<String> inconsistentShardIds,
-            final MultiStreamArgs multiStreamArgs) {
-        final Map<String, Lease> shardIdToNewLeaseMap = new HashMap<>();
-        final Map<String, Shard> shardIdToShardMapOfAllKinesisShards = constructShardIdToShardMap(shards);
+    static List<Shard> getShardListAtInitialPosition(@NonNull final ShardDetector shardDetector,
```

Comment:
Should this be a method of EmptyLeaseSync?

Reply:
Yeah we should be able to move this to emptyleasesync.

Reply:
In new revision I ended up removing the garbage collection from the interface, so we need this to stay in HierarchicalShardSyncer since garbage collection needs it.